contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,171 @@
+
+So, create a new `Topology` field in `Endpoints.Subsets.Addresses` for identifying what topology domain the endpoints pod exists, e.g. what host, rack, zone, region etc. In other words, copy the topology-related labels of node hosting the endpoint to `EndpointAddress.Topology`.
+
+```go
+type EndpointAddress struct {
+  // labels of node hosting the endpoint
+  Topology map[string]string
+}
+```
+
+## Endpoints Controller changes
+
+Endpoint Controller will populate the `Topology` for each `EndpointAddress`. We want `EndpointAddress.Topology` to tell the LB, such as kube-proxy what topology domain(e.g. host, rack, zone, region etc.) the endpoints is in.
+
+Endpoints controller will need to watch ALL nodes and maintain a key-value store for node cache - the key is the nodename and value is the node object. So, the new logic of endpoint controller might like:
+
+```go
+go watch service, pod, node
+for i, pod := range service backends; do
+  node := nodeCache[pod.Spec.NodeName]
+  // Copy all topology-related labels of node hosting endpoint to endpoint
+  // We can only include node labels referenced in the service spec's topology constraints
+  endpointAddress := &v1.EndpointAddress {}
+  for topologyKey in service.Spec; do
+    endpointAddress.Topology[topologyKey] = node.Labels[topologyKey]
+  done
+  endpoints.Subsets[i].Addresses = endpointAddress
+done
+```
+
+## Kube-proxy changes
+
+Kube-proxy will respect topology keys for each service, so kube-proxy on different nodes may create different proxy rules.
+
+Kube-proxy will watch its own node and will find the endpoints that are in the same topology domain as the node if `service.Topology.Mode != ignored`.
+
+The new logic of kube-proxy might like:
+
+```go
+go watch node with its nodename
+switch service.Topology.Mode {
+  case "ignored":
+    route request to an endpoint randomly
+  case "required":
+    endpointsMeetRequirement := make([]endpointInfo, 0)
+    topologyKey := service.Topology.Key
+    // filter out endpoints that does not meet the "hard" topology requirements
+    for i := range service's endpoints.Subsets; do
+      ss := endpoints.Subsets[i]
+      for j := range ss.Addresses; do
+        // check if endpoint are in the same topology domain as the node running kube-proxy
+        if ss.Addresses[j].Topology[topologyKey] == node.Labels[topologyKey]; then
+          endpointsMeetHardRequirement = append(endpointsMeetHardRequirement, endpoint)
+        fi
+      done
+    done
+    // If multiple endpoints match, randomly select one
+    if len(endpointsMeetHardRequirement) != 0; then 
+      route request to an endpoint in the endpointsMeetHardRequirement randomly
+    fi
+  case "preferred":
+    // Try to find endpoints that meet the "soft" topology requirements firstly,
+    // If no one match, randomly select one in service's endpoints.
```

Comment:
Maybe nitpicking but this struck me as wrong: kube-proxy should not pick one endpoint; it should tell the kernel all available endpoints and ask it to to route each request randomly to one of them.

Reply:
sure, I intend to do that. Thanks! And I did that "tell the kernel all available endpoints" in code PR: kubernetes/kubernetes#58732

Reply:
It's fixed now. Thanks!